### PR TITLE
Feature/decimals

### PR DESCRIPTION
**Pull request by vektor8891**
Implements: #10 

- Adding PLDecimal class to handle decimal numbers
- Handles scientific notation as well (e.g. `1.2E-23`)

### CODE REVIEW

**Comment by vektor8891** [comment]
@maxinteger i have created the class & related functions. can you pls help me how to wire them into the system?

**Review by maxinteger — CHANGES_REQUESTED** [review]
@vektor8891 nice PR

The only missing thing is the decimal function is not imported into `src/index.ts`
Also keep in mind PL not support decimal literals so the user must use a function call like `(decimal "123")` instead just type the numbers `123`. 
Make sure you are re exporting the `plDecimal` as `decimal` in `src/index.ts`

**Comment by vektor8891** [comment]
@maxinteger

1) I've added a function to auto simplify decimals to be in line with fraction implementation.

2) I would suggest to add support for decimals in PL. Having the program recognize decimals from input, or converting the class when needed would be very helpful for end-users. Otherwise the imprecise representation will still remain in the system (e.g. when dividing two integers, or converting fractions to decimals). What do you think?

**Comment by vektor8891** [comment]
@maxinteger What do you think if I merge the features of PLDecimal into PLNumber? In this way we don't need to define new literals.

File: src/data/decimal/decimalFn.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      decObj.decimals -= exponential
+    } else {
+      decObj.intValue *= Math.pow(10, exponential - decObj.decimals)
+      decObj.decimals = 0
+    }
+    return { intValue: decObj.intValue, decimals: decObj.decimals }
+  }
+}
+
+export function parseDecimalString(strValue: string): any {
```
Comment (maxinteger) [inline src/data/decimal/decimalFn.ts:42]: you can rewrite the function like this:
```
export function parseDecimalString(strValue: string): DecimalResult {
  assertNumeric(strValue)
  const [int, fraction] = strValue.split('.')
  const decimals = fraction?.length ?? 0
  const intValue = parseInt(int)
  return { intValue, decimals }
}
```

File: src/data/decimal/decimalFn.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  if (parts.length === 2 && assertNumeric(parts[0]) && assertNumeric(parts[1])) {
+    const intValueRounded = Math.abs(parseInt(parts[0]))
+    const exponential = parseFloat(parts[1])
+    return intValueRounded >= 1 && intValueRounded <= 9 && exponential === parseInt(parts[1])
+  } else {
+    return false
+  }
+}
+
+export function parseScientificString(strValue: string): any {
```
Comment (maxinteger) [inline src/data/decimal/decimalFn.ts:25]: Looks like you return with `{intValue, decimals}` objects few times, so you can create a helper interface for that fter the imports:

```
interface DecimalResult {
   intValue: number, 
   decimals: number
}

then you can use it instead `any`

File: src/data/decimal/decimalFn.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+export function simplifyDecimal(intValueOld: number, decimalsOld: number): any {
+  while (intValueOld % 10 === 0 && decimalsOld > 0) {
+    intValueOld /= 10
+    decimalsOld -= 1
+  }
+  return { intValue: intValueOld, decimals: decimalsOld }
+}
+
+export function expandDecimals(d1: PLDecimal, d2: PLDecimal): any {
```
Comment (maxinteger) [inline src/data/decimal/decimalFn.ts:63]: the return type is `{ maxDecimal: number, intValue1: number, intValue2: number }` instead of `any`

File: src/data/decimal/decimalFn.spec.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  getDecimalString,
+} from './decimalFn'
+import { PLDecimal } from './PLDecimal'
+
+const pld = plDecimal
+
+describe('stdlib/data/decimal/decimalFn', () => {
+  describe('plDecimal', () => {
+    it('should construct decimal', () => {
+      expect(pld('0.2')).toEqual(new PLDecimal(2, 1))
```
Comment (maxinteger) [inline src/data/decimal/decimalFn.spec.ts:19]: Testing is like playing chess with yourself, you must give your best to beat your implementation in the tests cases.
So you need to think about edge cases especially when you are doing string parsing where the input data set is much bigger then with numbers.

So please add some edge cases like `pld('hello world')` or `pld('0.2hello')` etc.
Because all our functions use the same parser you can do edge cases testings ones

**Review by maxinteger — CHANGES_REQUESTED** [review]

**Comment by vektor8891** [comment]
@maxinteger Thanks for your comments, I will have a look. What do you think of my suggestion of merging PLDecimal into PLNumber to avoid type collision during e.g. division?

**Comment by maxinteger** [comment]
@vektor8891 Actually we can replace the plNumber with the plDecimal implementation so it will be used everywhere

OR I can add a new number format to the language for decimal number ex: `123.42d` where the `d` postfix mean "decimal", in this case you can use easily the decimal numbers, but still need to convert manually between normal and decimals. For example `(add 1.5 1.5d)` will throw error and you need to do `(add (plDecimal 1.5) 1.5d)` or `(add 1.5 (plNumber 1.5d))` (this make more sense when you use variables)

I prefer the second option because it gives options for the develop which format s/he want to use, and also most of the languages have dedicated decimal number type

**Comment by vektor8891** [comment]
@maxinteger I see your point in keeping it separate in general. However, for Zsebtanár, I don't think there will be any use of regular floating points because they create more problems than they solve (by e.g. calculating faster). So my suggestion would be to merge PLDecimal into PLNumber, but only for Zsebtanár. I'm even open to remove PLDecimal from the stdlib altogether. What do you think?

**Comment by maxinteger** [comment]
My Original idea was to create a generic popuse language with Pocket lisp. This is why we are pushing all the Zsebtanár specific functions into the zsebtanar-proto repository.

But Decimal numbers are not Zsebtanár specific, so we can say PL support accurate number operations.

Please replace PLNumber implementation with PLDecimal

**Comment by vektor8891** [comment]
@maxinteger how does this change impact other numeric functions e.g. `pow`, `sqrt`, `sin`? do we have to rewrite them too?

File: src/data/decimal/decimalFn.ts (hunk 3 above)
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: fixed

File: src/data/decimal/decimalFn.spec.ts (hunk 4 above)
Comment (vektor8891) [reply src/data/decimal/decimalFn.spec.ts]: i added some checks for plNumber to handle edge cases

File: src/data/decimal/decimalFn.ts (hunk 2 above)
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: done

File: src/data/decimal/decimalFn.ts (hunk 1 above)
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: thanks for the idea! i updated the code (i did not change `intValue` as it consist the fraction part too):
```
export function parseDecimalString(strValue: string): DecimalResult {
  assertNumeric(strValue)
  const decimals = strValue.split('.').pop()?.length ?? 0
  const intValue = parseInt(strValue.replace('.', ''))
  return { intValue: intValue, decimals: decimals }
}
```

File: src/data/decimal/decimalFn.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import { assert } from '../../utils/assert'
+import { PLDecimal } from './PLDecimal'
+
+export function plDecimal(strValue: string): PLDecimal {
+  const decimalObj = parseNumString(strValue)
+  return new PLDecimal(decimalObj.intValue, decimalObj.decimals)
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function assertNumeric(strValue: any): any {
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: 1+3) The source of the function is [this](https://stackoverflow.com/questions/175739/built-in-way-in-javascript-to-check-if-a-string-is-a-valid-number#175787) link. As the comment says, `parseFloat` cannot check if the entire string is numeric (e.g. it can parse `0.2` from `0.2hello`). If you have better solution, please let me know.

2) I changed the return type to `boolean` as in the `assert` function.

File: src/data/decimal/decimalFn.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  assertNumeric(strValue)
+  const parts = strValue.split('.')
+  const decimals = parts.length === 1 ? 0 : parts[1].length
+  const intValue = parseInt(strValue.replace('.', ''))
+  return { intValue: intValue, decimals: decimals }
+}
+
+export function parseNumString(strValue: string): any {
+  assertNumeric(strValue)
+  return isScientific(strValue) ? parseScientificString(strValue) : parseDecimalString(strValue)
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: i modified the check to assertion in `parseScientificString`

File: src/data/decimal/decimalFn.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    decimalsOld -= 1
+  }
+  return { intValue: intValueOld, decimals: decimalsOld }
+}
+
+export function expandDecimals(d1: PLDecimal, d2: PLDecimal): any {
+  const maxDecimal = Math.max(d1.decimals, d2.decimals)
+  const intValue1 = d1.intValue * Math.pow(10, maxDecimal - d1.decimals)
+  const intValue2 = d2.intValue * Math.pow(10, maxDecimal - d2.decimals)
+  return { maxDecimal: maxDecimal, intValue1: intValue1, intValue2: intValue2 }
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: thanks! fixed it.

File: src/data/decimal/PLDecimal.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    Add<PLDecimal>,
+    Subtract<PLDecimal>,
+    Multiple<PLDecimal>,
+    Divide<PLDecimal>,
+    Negate<PLDecimal>,
+    PartialOrd<PLDecimal>,
+    Copy<PLDecimal> {
+  public static kind = 'Decimal'
+
+  private readonly _intValue: number
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/PLDecimal.ts]: i tried to be consistent with other existing classes (e.g. `PLFractionNumber`)

File: src/data/decimal/decimalFn.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function assertNumeric(strValue: any): any {
+  return assert(strValue === '' || isNaN(strValue), `Invalid decimal number: "${strValue}"`)
+}
+
+export function isScientific(strValue: string): boolean {
+  const parts = strValue.split(/[eE]/)
+  if (parts.length === 2 && assertNumeric(parts[0]) && assertNumeric(parts[1])) {
+    const intValueRounded = Math.abs(parseInt(parts[0]))
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: - It's ok to lose the fraction part. i only need to know if the number is between 1 and 9 (otherwise the number is not in a  scientific form)
- I would prefer not to use `parseFloat` because it is too smart: it can parse scientific numbers e.g. `1e3` even if that should already be excluded at `parts[0]`.

File: src/data/decimal/decimalFn.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  const decimalObj = parseNumString(strValue)
+  return new PLDecimal(decimalObj.intValue, decimalObj.decimals)
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function assertNumeric(strValue: any): any {
+  return assert(strValue === '' || isNaN(strValue), `Invalid decimal number: "${strValue}"`)
+}
+
+export function isScientific(strValue: string): boolean {
```
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: Thanks for the suggestion! i slightly modified it to `/^[-+]?[1-9](\.\d+)?[eE][-+]?\d+$/.test(strValue)`.

**Comment by vektor8891** [comment]
@maxinteger I've tried to address all your comments. Please let me know what you think.

File: src/utils/math.ts [hunk 11]
```diff
@@ -6,5 +6,9 @@ export const gcd = (a: number, b: number): number => {
 }
 
 export const isBelowEpsilon = (x: number): boolean => {
-  return Math.abs(x) < Number.EPSILON
+  return Math.abs(x) <= Number.EPSILON
+}
+
+export const floatEq = (a: number, b: number): boolean => {
```
Comment (maxinteger) [inline src/utils/math.ts:12]: Please add tests for this function too

File: src/data/number/numberFn.ts [hunk 12]
```diff
@@ -1,10 +1,90 @@
...
-import { plString } from '../string/stringFn'
 import { PLNumber } from './PLNumber'
 
-export const plNumber = (value: number): PLNumber => new PLNumber(value)
+export interface DecimalResult {
+  intValue: number
+  decimals: number
+}
+
+export function plFloat(x: number): PLNumber {
```
Comment (maxinteger) [inline src/data/number/numberFn.ts:9]: `plFloat` is looks like a new type and it is missleading for PL devs and for PL users as well.

I recommend to use the `plNumber` and overload it.

you need to refactore it a bit like this:
```
export function plNumber(value: string): PLNumber
export function plNumber(value: number, decimal: number): PLNumber
export function plNumber(value: number | string, decimal = 0): PLNumber {
  if (typeof value === 'string') {
  assertNumeric(strValue)
     const decimalObj = isScientific(strValue) ? parseScientificString(strValue) :  parseDecimalString(strValue)
     return plNumber(decimalObj.intValue, decimalObj.decimals)
  } else {
      assertInteger(intValue)
      assertInteger(decimals)
      return new PLNumber(intValue, decimals)
  }
}

File: src/data/number/numberFn.ts [hunk 13]
```diff
@@ -1,10 +1,90 @@
...
+}
+
+export function parseNumber(strValue: string): PLNumber {
+  assertNumeric(strValue)
+  const decimalObj = isScientific(strValue) ? parseScientificString(strValue) : parseDecimalString(strValue)
+  return plNumber(decimalObj.intValue, decimalObj.decimals)
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function assertNumeric(strValue: any): boolean {
```
Comment (maxinteger) [inline src/data/number/numberFn.ts:26]: please do not use `any` type, espeacially when the parameter name suggests `string` type

File: src/data/number/numberFn.ts [hunk 14]
```diff
@@ -1,10 +1,90 @@
...
+  return assert(strValue === '' || isNaN(strValue), `Invalid number: "${strValue}"`)
+}
+
+export const isRational = (x: number): boolean => {
+  const decimalStr = x.toString().split('.').pop()
+  const decimalLen = decimalStr ? decimalStr.length : 0
```
Comment (maxinteger) [inline src/data/number/numberFn.ts:32]: you can use a new and shorter syntax here: 
```
const decimalLen = decimalStr?.length ?? 0
```

File: src/data/number/numberFn.ts [hunk 15]
```diff
@@ -1,10 +1,90 @@
...
+  const isFraction = /\..*(\d)\1{12,}\d$/g.test(x.toString())
+  return decimalLen < 12 ? true : isFraction
```
Comment (maxinteger) [inline src/data/number/numberFn.ts:34]: why `12`?
If you must use a contant value, please crate a `const` outside of the this function and git it a proper name

File: src/data/number/numberFn.ts [hunk 16]
```diff
@@ -1,10 +1,90 @@
...
+
+export function parseNumber(strValue: string): PLNumber {
+  assertNumeric(strValue)
+  const decimalObj = isScientific(strValue) ? parseScientificString(strValue) : parseDecimalString(strValue)
+  return plNumber(decimalObj.intValue, decimalObj.decimals)
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function assertNumeric(strValue: any): boolean {
+  return assert(strValue === '' || isNaN(strValue), `Invalid number: "${strValue}"`)
```
Comment (maxinteger) [inline src/data/number/numberFn.ts:27]: string value never equel with `NaN`, if you really want to use this for check then you have to parse the value first:
```
isNaN(parseFloat(strValue))
```

**Review by maxinteger — CHANGES_REQUESTED** [review]
Unfortunately the code coverage with tests droped from 96% to 91%, you can see it under our conversation in github.
So please test all exported funtions, and let me know if you need more help to see how much of the code covered or not in local dev environment

File: src/data/decimal/decimalFn.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: I've rewritten the function using regexp.

File: src/data/decimal/decimalFn.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (vektor8891) [reply src/data/decimal/decimalFn.ts]: Another modification:
```
export function isScientific(strValue: string): boolean {
  if (strValue.startsWith('0')) {
    return /^0[eE]0$/.test(strValue)
  } else {
    return /^[-+]?[1-9](\.\d+)?[eE][-+]?\d+$/.test(strValue)
  }
}
```

File: src/utils/math.ts (hunk 11 above)
Comment (vektor8891) [reply src/utils/math.ts:12]: done.

File: src/data/number/numberFn.ts (hunk 12 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:9]: Done.

File: src/data/number/numberFn.ts (hunk 14 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:32]: i removed this function.

File: src/data/number/numberFn.ts (hunk 15 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:34]: I've removed this function.

File: src/data/number/numberFn.ts (hunk 16 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:27]: See my previous reply [here](https://github.com/maxinteger/pocket-lisp-stdlib/pull/14#discussion_r559078636).

File: src/data/number/numberFn.ts (hunk 13 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:26]: See my previous comment on the same issue [here](https://github.com/maxinteger/pocket-lisp-stdlib/pull/14#discussion_r559078636).

**Comment by vektor8891** [comment]
@maxinteger I've updated the code. Can you pls have a look?

File: src/data/number/numberFn.ts (hunk 13 above)
Comment (maxinteger) [reply src/data/number/numberFn.ts:26]: Sorry but your comment still not explain why do you use `any` type for `strValue`
Either the parameters name is wrong or the type of the parameter so you should
- use `string` type: `assertNumeric(strValue: string): boolean` OR
- use `unknown` type with proper parameter name: `assertNumeric(value: unknown): boolean`

File: src/data/number/numberFn.ts (hunk 16 above)
Comment (maxinteger) [reply src/data/number/numberFn.ts:27]: hmm I did not know about `isNaN` convert the passed value to Number. Good to know :D

File: src/data/number/numberFn.ts (hunk 13 above)
Comment (vektor8891) [reply src/data/number/numberFn.ts:26]: The problem is that `isNan` expects a number (even though I want to apply it to a string) so both `string` and `unknown` will give an error:

```
Argument of type 'string' is not assignable to parameter of type 'number'.
Argument of type 'unknown' is not assignable to parameter of type 'number'.
```

Comment (maxinteger) [reply src/data/number/numberFn.ts:26]: I see, you can do this then:

```
export function assertNumeric(strValue: string): boolean {
   return assert(strValue === '' || isNaN(Number(strValue)), `Invalid number: "${strValue}"`)
}
```

Comment (vektor8891) [reply src/data/number/numberFn.ts:26]: Thanks! It works now.

**Comment by vektor8891** [comment]
@maxinteger Thank you for your comments! I've updated the code.
